### PR TITLE
Add package ReFrame

### PR DESCRIPTION
**Pull request by victorusu**
This PR adds ReFrame to the list of supported applications.

ReFrame is a regression framework. More info can be found at the GitHub repo https://github.com/eth-cscs/reframe.

### CODE REVIEW

**Comment by alalazo** [comment]
@victorusu There are `flake8` issues. Can you fix them?

**Comment by vkarak** [comment]
@victorusu Can you please change `Reframe` to `ReFrame` in the title?

File: var/spack/repos/builtin/packages/reframe/package.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    focusing only on the functionality."""
+
+    homepage = 'https://reframe-hpc.readthedocs.io'
+    url      = 'https://github.com/eth-cscs/reframe/archive/v2.21.tar.gz'
+    git      = 'https://github.com/eth-cscs/reframe.git'
+
+    # notify when the package is updated.
+    maintainers = ['victorusu', 'vkarak']
+
+    version('develop',   branch='master')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/reframe/package.py]: I would call this master

File: var/spack/repos/builtin/packages/reframe/package.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    version('2.21',      sha256='f35d4fda2f9672c87d3ef664d9a2d6eb0c01c88218a31772a6645c32c8934c4d')
+    version('2.20',      sha256='310c18d705858bbe6bd9a2dc4d382b254c1f093b0671d72363f2111e8c162ba4')
+    version('2.17.3',    sha256='dc8dfb2ccb9a966303879b7cdcd188c47063e9b7999cbd5d6255223b066bf357')
+    version('2.17.2',    sha256='092241cdc15918040aacb922c806aecb59c5bdc3ff7db034a4f355d39aecc101')
+    version('2.17.1',    sha256='0b0d32a892607840a7d668f5dcea6f03f7022a26b23e5042a0faf5b8c41cb146')
+
+    depends_on('python@3.5:', type=('run'))
+    phases = ['install']
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/reframe/package.py]: This line isn’t needed

**Comment by alalazo** [comment]
@adamjstewart I noticed that the package is also distributed on PyPI as [reframe-hpc](https://pypi.org/project/ReFrame-HPC/). What's our policy on that? I think this package should `extend` Python and probably be called `py-*`, right?

**Comment by victorusu** [comment]
@alalazo, the project is called [ReFrame](https://github.com/eth-cscs/reframe). So, the name reframe is the one that the developers use. The `py-reframe-hpc` can be used as an alternative, but it is not the "real" project name, since the name ReFrame was already taken in pip repo. 

I vote against changing the name of the package.

**Comment by alalazo** [comment]
@victorusu Every Python package in Spack is prefixed by `py-` for instance:
```console
$ spack list numpy
==> 2 packages.
py-numpy  py-numpydoc
```
even if the project is called `NumPy`. Whether call it or not `py-reframe` vs. `py-reframe-hpc` can be discussed (considering that the other project seems abandoned and not very followed), but I think that at least this should `extend` Python and have the `py-` prefix.

@adamjstewart has surely more context than me on Python packages in general, so I'll leave the final decision up to him.

**Comment by victorusu** [comment]
@alalazo, the main difference between `ReFrame` and `NumPy` is that `ReFrame` is not a Python library. It is a standalone application implemented in Python.
Spack has other python applications that do not have the `py-` prefix in the package name...

Furthermore, why only for python applications (not the libraries) one needs to add the `py-` prefix?
Why not also add `cpp-`, `c-`, `ftn-` prefix for the other languages? This is very arbitrary...

If the question is about having `ReFrame` installed as a Python package as well, that's another PR. And that I easily do, too. But it is completely unrelated to the application ReFrame.

**Comment by adamjstewart** [comment]
See https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pythonpackage-vs-packages-that-use-python for my stance on naming/extends

**Comment by victorusu** [comment]
@adamjstewart, I am completely following the guidelines... 
> The thing these packages [busco, easybuild, ...] have in common is that they are command-line tools that just so happen to be written in Python

That's exactly ReFrame. 

Having said that, this ReFrame PR is not meant to use ReFrame as a library. And the guide line to install ReFrame is to just clone it and use a Python3.5+ interpreter...

**Comment by alalazo** [comment]
> Having said that, this ReFrame PR is not meant to use ReFrame as a library. And the guide line to install ReFrame is to just clone it and use a Python3.5+ interpreter...

My stance on that is that this is not that different from e.g. `tox` whose package is `py-tox`. Also, to be fair, from your `README` the first suggestion is:

> You may install ReFrame directly from PyPI through pip:
> ```
> pip install reframe-hpc
> ```

:wink: That said anything that is good for @adamjstewart is good for me.

**Comment by victorusu** [comment]
@alalazo, 👍

**Comment by vkarak** [comment]
> > Having said that, this ReFrame PR is not meant to use ReFrame as a library. And the guide line to install ReFrame is to just clone it and use a Python3.5+ interpreter...
> 
> My stance on that is that this is not that different from e.g. `tox` whose package is `py-tox`. Also, to be fair, from your `README` the first suggestion is:
> 
> > You may install ReFrame directly from PyPI through pip:
> > ```
> > pip install reframe-hpc
> > ```

If that part of the README in ReFrame is the problem and make people think that it is a Python library, I can take that and change it, although I believe it is not the case. The parallelism to `tox` imho is wrong as well. `tox` is a Python-specific tool and therefore I can understand it having the `py-` prefix. ReFrame is neither a Python tool nor a Python library. As in @adamjstewart's documentation, it is a tool that happens to be written in Python.

**Comment by scheibelp** [comment]
For now, `reframe` seems a suitable name for the Spack package

> the name ReFrame was already taken in pip repo.

In the future, it may be worth changing the name of this package to `reframe-hpc` for similar reasons (for now there are no packages conflicting).

**Comment by vkarak** [comment]
Thanks @scheibelp for merging.

**Pull request merged**